webhook/pr_review.go

```diff
@@ -0,0 +1,384 @@
+			}
+			position := getLastLinePosition(file.Patch)
+			err = submitReviewComment(owner, repo, prNumber, commitID, file.Filename, comment, position)
+			if err != nil {
+				log.Printf("Error submitting review comment for file %s: %v\n", file.Filename, err)
+			} else {
+				log.Printf("Review comment submitted for file %s\n\n", file.Filename)
+			}
+			log.Printf("Sleep %d seconds to avoid GitHub API rate limit\n", sleepTimeSecond)
+			time.Sleep(time.Duration(sleepTimeSecond) * time.Second)
+			builder.Reset()
+		}
+	}
+}
+
+func getChatGPTComment(patch string) (string, error) {
+	client, err := openai.New(
+		openai.WithToken(viper.GetString("openai.api_key")),
+		openai.WithModel(viper.GetString("openai.model")),
+		openai.WithOrgID(viper.GetString("openai.org_id")),
+		openai.WithProxyURL(viper.GetString("openai.proxy")),
+		openai.WithSocksURL(viper.GetString("openai.socks")),
+		openai.WithBaseURL(viper.GetString("openai.base_url")),
+		openai.WithTimeout(viper.GetDuration("openai.timeout")),
+		openai.WithMaxTokens(viper.GetInt("openai.max_tokens")),
+		openai.WithTemperature(float32(viper.GetFloat64("openai.temperature"))),
+	)
+	if err != nil {
+		log.Println("Error creating OpenAI client:", err)
+		return "", err
+	}
+
+	out, err := util.GetTemplateByString(
+		prompt.CodeReviewTemplate,
+		util.Data{
+			"file_diffs": patch,
+		},
+	)
+	if err != nil {
+		log.Println("Error creating prompt:", err)
+		return "", err
+	}
+
+	ctx := context.Background()
+	resp, err := client.Completion(ctx, out)
+	if err != nil {
+		log.Println("Error completing prompt:", err)
+		return "", err
+	}
+	summarizeMessage := resp.Content
+
+	log.Println("PromptTokens: " + strconv.Itoa(resp.Usage.PromptTokens) +
+		", CompletionTokens: " + strconv.Itoa(resp.Usage.CompletionTokens) +
+		", TotalTokens: " + strconv.Itoa(resp.Usage.TotalTokens))
+
+	// Output core review summary
+	log.Println("================Review Summary====================")
+	log.Println("\n" + strings.TrimSpace(summarizeMessage) + "\n\n")
+	log.Println("==================================================")
+
+	return summarizeMessage, nil
+}
+
+func getLastLinePosition(patch string) int {
+	lines := strings.Split(patch, "\n")
+	position := 0
+	for _, line := range lines {
+		if strings.HasPrefix(line, "@@") {
+			continue
+		}
+		position++
+	}
+	return position
+}
+
+func submitReviewComment(owner, repo string, prNumber int, commitID, filePath, comment string, position int) error {
+	client := &http.Client{}
+
+	commentRequest := CommentRequest{
+		Body:     comment,
+		Path:     filePath,
+		Position: position,
+		CommitID: commitID,
+	}
+
+	jsonData, err := json.Marshal(commentRequest)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest("POST", fmt.Sprintf(submitCommentsURL, apiBaseurl, owner, repo, prNumber), bytes.NewBuffer(jsonData))
+	if err != nil {
+		log.Println("Error creating request:", err)
```

Comment:
Overall, the code appears to be well-structured and readable. However, there are a few areas where improvements could be made:

1. Error handling: In some cases, the code simply returns an empty string if an error occurs. This makes it difficult to determine the root cause of the error, and could lead to unexpected behavior downstream. It would be better to return a meaningful error message, or to log the error and continue execution if possible.

2. Security: The code appears to be making HTTP requests without verifying the server's identity using SSL/TLS. This could leave the application vulnerable to man-in-the-middle attacks. It would be better to use HTTPS for all requests, and to validate the server's certificate.

3. Code quality: The code could benefit from more comments and documentation. This would make it easier for others to understand what the code does and how it works.

4. Performance: The code sleeps for a fixed amount of time after each API call, which could lead to longer execution times than necessary. It would be better to use an exponential backoff algorithm or to use a rate-limiting library to avoid hitting API rate limits.

5. Test coverage: The code does not appear to have any unit tests. Adding tests would make it easier to ensure that the code works as expected and to catch regressions.